From 49e6c01bbb80acf733484b56d50edea3d97220bf Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 11:44:21 -0400
Subject: [PATCH] add a SLAB_HARDENED configuration option

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 init/Kconfig | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index ab2486a41037..49309b4a8eec 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2009,6 +2009,14 @@ config SLAB_FREELIST_HARDENED
 	  sacrifies to harden the kernel slab allocator against common
 	  freelist exploit methods.
 
+config SLAB_HARDENED
+	default y
+	depends on SLUB
+	bool "Hardened SLAB infrastructure"
+	help
+          Make minor performance sacrifices to harden the kernel slab
+          allocator.
+
 config SLUB_CPU_PARTIAL
 	default y
 	depends on SLUB && SMP
