From 868ad41c67b1d3cf466a211e238b800c2babefa0 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 28 May 2017 18:47:47 -0400
Subject: [PATCH] mark size_index as __ro_after_init

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
---
 mm/slab_common.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/slab_common.c b/mm/slab_common.c
index 2faad867f636..65e6a0c7c6f9 100644
--- a/mm/slab_common.c
+++ b/mm/slab_common.c
@@ -844,7 +844,7 @@ EXPORT_SYMBOL(kmalloc_dma_caches);
  * of two cache sizes there. The size of larger slabs can be determined using
  * fls.
  */
-static s8 size_index[24] = {
+static s8 size_index[24] __ro_after_init = {
 	3,	/* 8 */
 	4,	/* 16 */
 	5,	/* 24 */
