From aa686a1e7f13430a76bf1d3411bc7efe28f09519 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 16 Aug 2017 14:14:00 -0400
Subject: [PATCH] use kernel command-line for early entropy

---
 init/main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/main.c b/init/main.c
index c0f8b335f33..fb6e10d7498 100644
--- a/init/main.c
+++ b/init/main.c
@@ -498,6 +498,7 @@ asmlinkage void __init start_kernel(void)
 	page_address_init();
 	pr_notice("%s", linux_banner);
 	setup_arch(&command_line);
+	add_device_randomness(command_line, strlen(command_line));
 	/*
 	 * Set up the the initial canary ASAP:
 	 */
