From a89331041a9e0d3b2ae9609e130a64cb14b2a44f Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 7 May 2017 00:28:23 -0400
Subject: [PATCH] add __read_only for non-init related usage

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
---
 include/linux/cache.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/cache.h b/include/linux/cache.h
index d742c57eaee5..f0222c070458 100644
--- a/include/linux/cache.h
+++ b/include/linux/cache.h
@@ -37,6 +37,8 @@
 #define __ro_after_init __section(".data..ro_after_init")
 #endif
 
+#define __read_only __ro_after_init
+
 #ifndef ____cacheline_aligned
 #define ____cacheline_aligned __attribute__((__aligned__(SMP_CACHE_BYTES)))
 #endif
