From 3aa5cb318bc1f9a2256691a8a788a0b501fb8206 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 28 May 2017 18:51:30 -0400
Subject: [PATCH] mark kmem_cache as __ro_after_init

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
---
 mm/slab_common.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/slab_common.c b/mm/slab_common.c
index 2e23ad39dd6c..fa1c9cbf1dbf 100644
--- a/mm/slab_common.c
+++ b/mm/slab_common.c
@@ -36,7 +36,7 @@
 enum slab_state slab_state __ro_after_init;
 LIST_HEAD(slab_caches);
 DEFINE_MUTEX(slab_mutex);
-struct kmem_cache *kmem_cache;
+struct kmem_cache *kmem_cache __ro_after_init;
 
 #ifdef CONFIG_HARDENED_USERCOPY
 bool usercopy_fallback __ro_after_init =
