diff --git a/Makefile b/Makefile
index 006700fbb6525..7021aa85afd1e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 VERSION = 5
 PATCHLEVEL = 10
-SUBLEVEL = 195
+SUBLEVEL = 196
 EXTRAVERSION =
 NAME = Dare mighty things
 
diff --git a/fs/configfs/dir.c b/fs/configfs/dir.c
index 0b7e9ab517d58..12388ed4faa59 100644
--- a/fs/configfs/dir.c
+++ b/fs/configfs/dir.c
@@ -479,7 +479,6 @@ static struct dentry * configfs_lookup(struct inode *dir,
 	if (!configfs_dirent_is_ready(parent_sd))
 		goto out;
 
-	spin_lock(&configfs_dirent_lock);
 	list_for_each_entry(sd, &parent_sd->s_children, s_sibling) {
 		if (sd->s_type & CONFIGFS_NOT_PINNED) {
 			const unsigned char * name = configfs_get_name(sd);
@@ -492,7 +491,6 @@ static struct dentry * configfs_lookup(struct inode *dir,
 			break;
 		}
 	}
-	spin_unlock(&configfs_dirent_lock);
 
 	if (!found) {
 		/*
