From 52626f3c20ec8f22422b1ee1630d0fd8eb3dac6d Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 10 Jan 2017 20:47:23 -0500
Subject: [PATCH] disable slab merging by default

---
 mm/slub.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/slub.c b/mm/slub.c
index a4cba911731..7430d06bc70 100644
--- a/mm/slub.c
+++ b/mm/slub.c
@@ -2685,7 +2685,7 @@ static int slub_min_objects;
  * Merge control. If this is set then no merging of slab caches will occur.
  * (Could be removed. This was introduced to pacify the merge skeptics.)
  */
-static int slub_nomerge;
+static int slub_nomerge = 1;
 
 /*
  * Calculate the order of allocation given an slab object size.
