From 0ca06f17d39c020083c480dcb07f8fbad924f229 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 11:38:54 -0400
Subject: [PATCH] add missing cache_from_obj !PageSlab check

Taken from PaX.

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
---
 mm/slab.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/slab.h b/mm/slab.h
index 803b124bb085..c037436d8794 100644
--- a/mm/slab.h
+++ b/mm/slab.h
@@ -371,6 +371,9 @@ static inline struct kmem_cache *cache_from_obj(struct kmem_cache *s, void *x)
 		return s;
 
 	page = virt_to_head_page(x);
+#ifdef CONFIG_SLAB_HARDENED
+	BUG_ON(!PageSlab(page));
+#endif
 	cachep = page->slab_cache;
 	if (slab_equal_or_root(cachep, s))
 		return cachep;
