From b3bb80e674d10c2d02f2b3bceef167e893116c87 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 30 Mar 2017 23:05:53 -0400
Subject: [PATCH] real slab_equal_or_root check for !MEMCG_KMEM

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
Signed-off-by: randomhydrosol <randomhydrosol@glassrom.org>
---
 mm/slab.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/mm/slab.h b/mm/slab.h
index 598bc0a769ce..d430e6a3d623 100644
--- a/mm/slab.h
+++ b/mm/slab.h
@@ -314,7 +314,11 @@ static inline bool is_root_cache(struct kmem_cache *s)
 static inline bool slab_equal_or_root(struct kmem_cache *s,
 				      struct kmem_cache *p)
 {
+#ifdef CONFIG_SLAB_HARDENED
+	return p == s;
+#else
 	return true;
+#endif
 }
 
 static inline const char *cache_name(struct kmem_cache *s)
