From 4ba727c32a9731dc684ff9969dd8015ad18528d4 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 11:44:21 -0400
Subject: [PATCH] add a SLAB_HARDENED configuration option

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>
Signed-off-by: randomhydrosol <randomhydrosol@glassrom.org>
---
 init/Kconfig | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index ad25e2be4b5b..547b09d22b76 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1759,6 +1759,14 @@ config SLAB_FREELIST_HARDENED
 	  sacrifies to harden the kernel slab allocator against common
 	  freelist exploit methods.
 
+config SLAB_HARDENED
+	default y
+	depends on SLUB
+	bool "Hardened SLAB infrastructure"
+	help
+          Make minor performance sacrifices to harden the kernel slab
+          allocator.
+
 config SLUB_CPU_PARTIAL
 	default y
 	depends on SLUB && SMP
